File: Dictionaries and Sets/1679. Max Number of K-Sum Pairs/1679. Max Number of K-Sum Pairs.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxOperations = function (nums, k) {
  let c = {};
  let ops = 0;

  for (const n of nums) {
    const compliment = k - n;
    if (c[compliment] > 0) {
      c[compliment]--;
      ops++;
    } else c[n] = (c[n] || 0) + 1;
  }

  return ops;
};

// Time: O(n)
// Space: O(n)

File: Dictionaries and Sets/1679. Max Number of K-Sum Pairs/1679. Max Number of K-Sum Pairs.py
class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        c = defaultdict(int)
        ops = 0

        for n in nums:
            compliment = k - n
            if c[compliment] > 0:
                c[compliment] -= 1
                ops += 1
            else:
                c[n] += 1
        
        return ops

# Time: O(n)
# Space: O(n)
